src/Compilers/Core/Portable/Compilation/Compilation.cs

```diff
@@ -2449,99 +2449,36 @@ internal void EnsureAnonymousTypeTemplates(CancellationToken cancellationToken)
-                    if (!metadataOnly && IsRealSigned && this.Options.StrongNameProvider.Capability == SigningCapability.SignsStream)
-                    {
-                        Debug.Assert(Options.StrongNameProvider != null);
-
-                        try
-                        {
-                            signingInputStream = Options.StrongNameProvider.CreateInputStream();
-                        }
-                        catch (IOException e)
-                        {
-                            throw new Cci.PeWritingException(e);
-                        }
-
-                        retStream = signingInputStream;
-                    }
-                    else
-                    {
-                        signingInputStream = null;
-                        retStream = peStream;
-                    }
-
-                    return retStream;
+                    Stream ret;
+                    (peStream, signingInputStream, ret) = GetPeStream(metadataDiagnostics, peStreamProvider, metadataOnly);
+                    return ret;
                 };
 
-                Func<Stream> getRefPeStream;
-                if (metadataPEStreamProvider != null)
+                Func<Stream> getRefPeStream = () =>
                 {
-                    getRefPeStream = () =>
-                    {
-                        if (metadataDiagnostics.HasAnyErrors())
-                        {
-                            return null;
-                        }
+                    refPeStream = GetRefPeStream(metadataDiagnostics, metadataPEStreamProvider);
+                    return refPeStream;
+                };
 
-                        refPeStream = metadataPEStreamProvider.GetOrCreateStream(metadataDiagnostics);
-                        Debug.Assert(refPeStream != null || metadataDiagnostics.HasAnyErrors());
-                        return refPeStream;
-                    };
-                }
-                else
+                var includePortablePdbStream = moduleBeingBuilt.DebugInformationFormat == DebugInformationFormat.PortablePdb && pdbStreamProvider != null;
+                Func<Stream> getPortablePdbStream = () =>
                 {
-                    getRefPeStream = null;
-                }
+                    portablePdbStream = GetPortablePdbStream(metadataDiagnostics, pdbStreamProvider);
+                    return portablePdbStream;
+                };
 
                 try
                 {
                     if (SerializePeToStream(
                         moduleBeingBuilt,
                         metadataDiagnostics,
-                        this.MessageProvider,
+                        MessageProvider,
                         getPeStream,
-                        getRefPeStream,
-                        getPortablePdbStream,
+                        metadataPEStreamProvider != null ? getRefPeStream : null,
+                        includePortablePdbStream ? getPortablePdbStream : null,
```

Comment:
Consider inlining all three local functions now that each function is  essentially one line. Inlining will make it clear which locals are involved, and should allow several locals to be removed.

Reply:
What do you mean by "inlining" in this context?

Reply:
Sorry, by inlining I mean replacing the `Func<Stream>` locals with the actual lambdas in the call to `SerializePeToStream`, and removing any `stream` locals that are not needed. I'd overlooked the fact that `signingInputStream` and `peStream` are used later in the method however.